text/0031-transactions-with-contingent-inputs.md

```diff
@@ -0,0 +1,191 @@
+  Because the `RingMLSAG` signs over the `ContingentInputRules`, these rules cannot be tampered with once the `RingMLSAG` is signed.
+
+A "signed contingent input" is a payload consisting of:
+* A `TxIn`, with `ContingentInputRules`.
+* A `pseudo_output_commitment` for this ring.
+* A `RingMLSAG` signing this.
+* A token id, value, and blinding factor for the `pseudo_output_commitment`.
+* A token id, value, and blinding factor for any output in the `ContingentInputRules`.
+
+Anyone who owns a `TxOut` can use it to create a `SignedContingentInput`. The signed contingent input can be thought of as an offer
+to trade "this for that" -- you are willing to give away your `TxOut` if someone produces for you the `TxOut` that you specify.
+Once you sign the contingent input, you can give it away, and anyone who has it can fulfill your request to trade.
+
+Giving away the signed contingent input does not reveal your public address, nor which `TxOut` in the ledger you actually own.
+It does reveal the token id and value of your pseudo-output, but this is unavoidable in any such scheme.
+
+Anyone who receives a `SignedContingentInput` can simply add it to the transaction builder and incorporate both the input and the required
+outputs into a transaction. As long as they produce a balanced transaction at the end, their transaction will be valid and they will be able
+to perform the swap with their counterparty atomically in a normal transaction.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+TODO more detail
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+The main drawback is that it increases the complexity of transaction validation.
+If a simpler or better way of doing swaps is discovered and this is used only rarely, we will still have to carry the maintanence burden forwards.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+> - Why is this design the best in the space of possible designs?
+
+We can measure the complexity of a swap protocol by considering the number of rounds of interactions
+between the participants and the number of on-chain events. In this proposal, these are both 1,
+which is optimal.
+
+Privacy is also a key consideration. In this proposal, the first party sends only a `SignedContingentInput`,
+so the second party does not learn their public address, only what they propose to trade and for how much.
+The first party does not learn anything about the second participant except that they executed the trade.
+The second party is unable to send subsequent transactions to the first party, since all they got was an output
+created by the first party, and the TxOut `public_key` is enforced to be unique across the ledger.
+
+In a realistic situation, it seems likely that the first and second parties know who each other are, since
+they must initiate the interaction somehow. However, that is not required by the protocol proposed here.
```

Comment:
Do you imagine some evolution where there could be "contingent input pools?" Not necessarily on MobileCoin's layer1 quite yet, but this seems like a great use case

Reply:
Yeah it seems quite plausible to me, like a `SignedContingentInput` is effectively like a limit order in an order book I think, except that there's no custodian and anyone who has the order is capable of executing it if they want. So you could conceivably have like a distributed order book of inputs like this, transmitted via a gossip network or something, and any participant is capable of executing any of the orders.

I think it would also be possible to merge lots of orders into one transaction and execute them all at once.